keps/sig-apps/4368-support-managed-by-label-for-batch-jobs/README.md

```diff
@@ -0,0 +1,1098 @@
+(see [implementation overview](#implementation-overview)).
+
+Additionally, re-evaluate extending the `kubectl` command-line tool
+before [GA](#ga). We could extend the command to provide useful debugging
+information with the following:
+- new `MANAGED_BY` column for `kubectl get job -owide` (possibly also without `-owide`)
+- a line in the `kubectl describe job` output, just before the list of events,
+providing a user readable information if the Job is synchronized by a custom
+controller.
+
+Alternative ideas considered were
+[a dedicated condition](#condition-to-indicated-job-is-skipped)
+and [events](#event-indicating-the-job-is-skipped).
+
+#### Custom controllers not compatible with API assumptions by CronJob
+
+Currently, the validation of the Job status API is rather relaxed, allowing for
+potential situations, when modified by custom controllers, which might be
+problematic for the API consumers, like the built-in CronJob controller.
+
+In particular, the CronJob relies on the assumption that the `CompletionTime`
+is only set to successful jobs. Also, that a job does not flip its state from
+Complete to Failed (or the other way round). Also, a finished job should not
+flip back to non-finished.
+
+In order to mitigate for this risk, we strengthen the validation of the Job
+Status API to prevent faulty custom controllers to do unexpected changes to the
+status (see [here](#job-status-validation)).
+
+
+## Design Details
+
+#### API
+
+```golang
+const (
+  ...
+	// LabelManagedBy is used to indicate the controller or entity that manages
+	// an Job.
+	LabelManagedBy = "batch.kubernetes.io/managed-by"
+)
+```
+
+#### Implementation overview
+
+We skip synchronization of the Jobs with the `managed-by` label, if it has any
+different value than `job-controller.k8s.io`. When the synchronization is skipped,
+the name of the controller managing the Job object is logged.
+
+We leave the particular place at which the synchronization is skipped as
+implementation detail which can be determined during the implementation phase,
+however, two candidate places are:
+1. inside `syncJob` function
+2. inside `enqueueSyncJobInternal` function
+
+Note that, if we skip inside `enqueueSyncJobInternal` we may save on some memory
+needed to needlessly enqueue the Job keys.
+
+There is no validation for the values of the field beyond that of standard
+permitted label values.
+
+#### Job status validation
+
+We will add the following validation rules, which are currently satisfied by the
+built-in Job controller, to mitigate the risk of [here](#custom-controllers-not-compatible-with-api-assumptions-by-cronjob):
+- the `CompletionTime` field is only set for jobs with `Failed` condition (per
+  [API comment](https://github.com/kubernetes/kubernetes/blob/48b68582b89b0ae9ad4d435516b2dd5943f48cd3/pkg/apis/batch/types.go#L436)).
+- once the `Failed=True` or `Complete=True` conditions are added, their status
+  cannot be changed. They are also mutually exclusive.
```

Comment:
Also:
The format of completedIndexes and failedIndexes. Maybe even that indexes have to be within the range of completions.

Reply:
Added this point